File: avalon-react-10.1.0/services/pagoService.ts
// services/pagoService.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8083/api';

export interface PagoRequest {
    ventaId: number;
    metodoPago: 'TRANSFERENCIA' | 'TARJETA' | 'DEUNA';
    comprobanteUrl?: string;
}

export interface PagoResponse {
    id: number;
    ventaId: number;
    numeroFactura: string;
    monto: number;
    metodoPago: string;
    estadoPago: string;
    referenciaTransaccion?: string;
    comprobanteUrl?: string;
    fechaPago?: string;
    fechaCreacion: string;
    fechaActualizacion: string;
    clienteId: number;
    clienteNombre: string;
    emprendedorId: number;
}

export interface AprobarPagoRequest {
    emprendedorId: number;
    observaciones?: string;
}

export const crearPago = async (pago: PagoRequest): Promise<PagoResponse> => {
    const response = await fetch(`${API_BASE_URL}/pagos`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(pago),
    });

    if (!response.ok) {
        throw new Error('Error al crear el pago');
    }

    return response.json();
};

export const obtenerPagoPorVenta = async (ventaId: number): Promise<PagoResponse> => {
    const response = await fetch(`${API_BASE_URL}/pagos/venta/${ventaId}`);
    
    if (!response.ok) {
        throw new Error('Error al obtener el pago');
    }

    return response.json();
};

export const aprobarPago = async (pagoId: number, request: AprobarPagoRequest): Promise<PagoResponse> => {
    const response = await fetch(`${API_BASE_URL}/pagos/${pagoId}/aprobar`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        throw new Error('Error al aprobar el pago');
    }

    return response.json();
};

export const rechazarPago = async (pagoId: number, request: AprobarPagoRequest): Promise<PagoResponse> => {
    const response = await fetch(`${API_BASE_URL}/pagos/${pagoId}/rechazar`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
    });

    if (!response.ok) {
        throw new Error('Error al rechazar el pago');
    }

    return response.json();
};

export const subirComprobante = async (pagoId: number, comprobanteUrl: string): Promise<PagoResponse> => {
    const response = await fetch(`${API_BASE_URL}/pagos/${pagoId}/comprobante`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ comprobanteUrl }),
    });

    if (!response.ok) {
        throw new Error('Error al subir el comprobante');
    }

    return response.json();
};

export const listarPagosPendientes = async (emprendedorId: number): Promise<PagoResponse[]> => {
    const response = await fetch(`${API_BASE_URL}/pagos/emprendedor/${emprendedorId}/pendientes`);
    
    if (!response.ok) {
        throw new Error('Error al listar pagos pendientes');
    }

    return response.json();
};

export const listarPagosPorEmprendedor = async (emprendedorId: number): Promise<PagoResponse[]> => {
    const response = await fetch(`${API_BASE_URL}/pagos/emprendedor/${emprendedorId}`);
    
    if (!response.ok) {
        throw new Error('Error al listar pagos del emprendedor');
    }

    return response.json();
};

export const procesarPagoSimulado = async (metodoPago: 'TARJETA' | 'DEUNA'): Promise<{
    success: boolean;
    transactionId: string;
    message: string;
}> => {
    console.log(`🎭 Simulando pago con ${metodoPago}...`);
    
    // Simular delay de procesamiento
    await new Promise(resolve => setTimeout(resolve, 1500));
    
    const mockTransactionId = `${metodoPago}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    
    return {
        success: true,
        transactionId: mockTransactionId,
        message: `Pago con ${metodoPago} procesado exitosamente (simulado)`
    };
};

export default {
    crearPago,
    obtenerPagoPorVenta,
    aprobarPago,
    rechazarPago,
    subirComprobante,
    listarPagosPendientes,
    listarPagosPorEmprendedor,
    procesarPagoSimulado
};

File: Microservicios/msvc-ventas/src/main/java/com/example/msvc_ventas/domain/service/PagoService.java
package com.example.msvc_ventas.domain.service;

import com.example.msvc_ventas.domain.model.SeguimientoLogistica;
import com.example.msvc_ventas.infrastructure.persistence.entity.PagoEntity;
import com.example.msvc_ventas.infrastructure.persistence.entity.VentaEntity;
import com.example.msvc_ventas.infrastructure.persistence.repository.PagoRepository;
import com.example.msvc_ventas.infrastructure.persistence.repository.VentaJpaRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class PagoService {

    private final PagoRepository pagoRepository;
    private final VentaJpaRepository ventaJpaRepository;
    private final SeguimientoLogisticaService seguimientoLogisticaService;
    private final FacturaService facturaService;

    /**
     * Crear un nuevo pago asociado a una venta
     * - TRANSFERENCIA/DEUNA: Estado PENDIENTE (requiere aprobación manual)
     * - TARJETA: Estado APROBADO automático (procesado por pasarela)
     */
    @Transactional
    public PagoEntity crearPago(Long ventaId, PagoEntity.MetodoPago metodoPago, String comprobanteUrl) {
        log.info("Creando pago para venta ID: {} con método: {}", ventaId, metodoPago);
        
        VentaEntity venta = ventaJpaRepository.findById(ventaId)
                .orElseThrow(() -> new IllegalArgumentException("Venta no encontrada: " + ventaId));

        // Determinar estado inicial según método de pago
        PagoEntity.EstadoPago estadoInicial = (metodoPago == PagoEntity.MetodoPago.TARJETA) 
                ? PagoEntity.EstadoPago.APROBADO 
                : PagoEntity.EstadoPago.PENDIENTE;

        PagoEntity pago = PagoEntity.builder()
                .venta(venta)
                .monto(venta.getTotal())
                .metodoPago(metodoPago)
                .estadoPago(estadoInicial)
                .comprobanteUrl(comprobanteUrl)
                .fechaCreacion(LocalDateTime.now())
                .fechaActualizacion(LocalDateTime.now())
                .build();

        // Si es pago con tarjeta, establecer fecha de pago inmediata
        if (metodoPago == PagoEntity.MetodoPago.TARJETA) {
            pago.setFechaPago(LocalDateTime.now());
        }

        PagoEntity pagoGuardado = pagoRepository.save(pago);
        log.info("Pago creado con ID: {} en estado {}", pagoGuardado.getId(), estadoInicial);
        
        // Si es pago con tarjeta, actualizar venta a COMPLETADA y APROBADO automáticamente
        if (metodoPago == PagoEntity.MetodoPago.TARJETA) {
            venta.setEstado(VentaEntity.EstadoVenta.COMPLETADA);
            venta.setEstadoPago(VentaEntity.EstadoPago.APROBADO);
            venta.setFechaActualizacion(LocalDateTime.now());
            ventaJpaRepository.save(venta);
            
            // Inicializar seguimiento logístico automáticamente
            try {
                seguimientoLogisticaService.inicializarSeguimientoPorPagoAprobado(
                        venta.getId(), 
                        "Pago automático con tarjeta"
                );
                log.info("Seguimiento logístico inicializado automáticamente para venta ID: {}", venta.getId());
            } catch (Exception e) {
                log.error("Error al inicializar seguimiento logístico para venta ID: {}", venta.getId(), e);
            }
        }
        
        return pagoGuardado;
    }

    /**
     * Aprobar un pago (emprendedor)
     */
    @Transactional
    public PagoEntity aprobarPago(Long pagoId, Long emprendedorId) {
        log.info("Aprobando pago ID: {} por emprendedor ID: {}", pagoId, emprendedorId);
        
        PagoEntity pago = pagoRepository.findById(pagoId)
                .orElseThrow(() -> new IllegalArgumentException("Pago no encontrado: " + pagoId));

        // Verificar que el pago pertenece al emprendedor
        if (!pago.getVenta().getEmprendedorId().equals(emprendedorId)) {
            throw new IllegalStateException("El pago no pertenece a este emprendedor");
        }

        if (pago.getEstadoPago() != PagoEntity.EstadoPago.PENDIENTE) {
            throw new IllegalStateException("Solo se pueden aprobar pagos en estado PENDIENTE");
        }

        pago.setEstadoPago(PagoEntity.EstadoPago.APROBADO);
        pago.setFechaPago(LocalDateTime.now());
        pago.setFechaActualizacion(LocalDateTime.now());

        PagoEntity pagoActualizado = pagoRepository.save(pago);
        
        // Actualizar estado de la venta
        VentaEntity venta = pago.getVenta();
        venta.setEstado(VentaEntity.EstadoVenta.COMPLETADA);
        venta.setEstadoPago(VentaEntity.EstadoPago.APROBADO);
        venta.setFechaActualizacion(LocalDateTime.now());
        ventaJpaRepository.save(venta);
        
        // 🚀 INICIALIZAR SEGUIMIENTO LOGÍSTICO
        try {
            seguimientoLogisticaService.inicializarSeguimientoPorPagoAprobado(
                    venta.getId(), 
                    "Emprendedor ID: " + emprendedorId
            );
            log.info("Seguimiento logístico inicializado para venta ID: {}", venta.getId());
        } catch (Exception e) {
            log.error("Error al inicializar seguimiento logístico para venta ID: {}", venta.getId(), e);
            // No fallar la transacción si falla el seguimiento
        }
        
        // 🧾 GENERAR FACTURA PDF DE FORMA ASÍNCRONA (no bloquea la respuesta)
        final Long ventaIdFinal = venta.getId();
        new Thread(() -> {
            try {
                Thread.sleep(100); // Pequeña pausa para asegurar que la transacción se complete
                String rutaFactura = facturaService.generarFacturaPDF(ventaIdFinal);
                log.info("✅ Factura generada en segundo plano: {}", rutaFactura);
            } catch (Exception e) {
                log.error("⚠️ Error al generar factura PDF para venta ID: {}", ventaIdFinal, e);
            }
        }).start();
        
        log.info("Pago ID: {} aprobado exitosamente", pagoId);
        return pagoActualizado;
    }

    /**
     * Rechazar un pago (emprendedor)
     */
    @Transactional
    public PagoEntity rechazarPago(Long pagoId, Long emprendedorId) {
        log.info("Rechazando pago ID: {} por emprendedor ID: {}", pagoId, emprendedorId);
        
        PagoEntity pago = pagoRepository.findById(pagoId)
                .orElseThrow(() -> new IllegalArgumentException("Pago no encontrado: " + pagoId));

        // Verificar que el pago pertenece al emprendedor
        if (!pago.getVenta().getEmprendedorId().equals(emprendedorId)) {
            throw new IllegalStateException("El pago no pertenece a este emprendedor");
        }

        if (pago.getEstadoPago() != PagoEntity.EstadoPago.PENDIENTE) {
            throw new IllegalStateException("Solo se pueden rechazar pagos en estado PENDIENTE");
        }

        pago.setEstadoPago(PagoEntity.EstadoPago.RECHAZADO);
        pago.setFechaActualizacion(LocalDateTime.now());

        PagoEntity pagoActualizado = pagoRepository.save(pago);
        
        // Actualizar estado de la venta
        VentaEntity venta = pago.getVenta();
        venta.setEstado(VentaEntity.EstadoVenta.CANCELADA);
        venta.setFechaActualizacion(LocalDateTime.now());
        ventaJpaRepository.save(venta);
        
        log.info("Pago ID: {} rechazado", pagoId);
        return pagoActualizado;
    }

    /**
     * Subir comprobante de transferencia
     */
    @Transactional
    public PagoEntity subirComprobante(Long pagoId, String comprobanteUrl) {
        log.info("Subiendo comprobante para pago ID: {}", pagoId);
        
        PagoEntity pago = pagoRepository.findById(pagoId)
                .orElseThrow(() -> new IllegalArgumentException("Pago no encontrado: " + pagoId));

        if (pago.getMetodoPago() != PagoEntity.MetodoPago.TRANSFERENCIA) {
            throw new IllegalStateException("Solo se puede subir comprobante para pagos por transferencia");
        }

        pago.setComprobanteUrl(comprobanteUrl);
        pago.setFechaActualizacion(LocalDateTime.now());

        return pagoRepository.save(pago);
    }

    /**
     * Obtener pago por ID
     */
    public PagoEntity obtenerPagoPorId(Long id) {
        return pagoRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Pago no encontrado: " + id));
    }

    /**
     * Obtener pago por venta ID
     */
    public PagoEntity obtenerPagoPorVentaId(Long ventaId) {
        return pagoRepository.findByVentaId(ventaId)
                .orElseThrow(() -> new IllegalArgumentException("No existe pago para la venta: " + ventaId));
    }

    /**
     * Listar pagos pendientes de un emprendedor
     */
    public List<PagoEntity> listarPagosPendientesPorEmprendedor(Long emprendedorId) {
        return pagoRepository.findByEmprendedorIdAndEstadoPago(emprendedorId, PagoEntity.EstadoPago.PENDIENTE);
    }

    /**
     * Listar todos los pagos de un emprendedor
     */
    public List<PagoEntity> listarPagosPorEmprendedor(Long emprendedorId) {
        return pagoRepository.findByEmprendedorId(emprendedorId);
    }

    /**
     * Listar todos los pagos por estado
     */
    public List<PagoEntity> listarPagosPorEstado(PagoEntity.EstadoPago estado) {
        return pagoRepository.findByEstadoPago(estado);
    }

    /**
     * Procesar pago con tarjeta (Stripe/pasarela de pagos)
     * Este método es específico para pagos ya procesados por la pasarela
     */
    @Transactional
    public PagoEntity procesarPagoTarjeta(Long ventaId, String referenciaTransaccion, String comprobanteUrl) {
        log.info("Procesando pago con tarjeta para venta ID: {}, referencia: {}", ventaId, referenciaTransaccion);
        
        VentaEntity venta = ventaJpaRepository.findById(ventaId)
                .orElseThrow(() -> new IllegalArgumentException("Venta no encontrada: " + ventaId));

        // Crear pago directamente APROBADO
        PagoEntity pago = PagoEntity.builder()
                .venta(venta)
                .monto(venta.getTotal())
                .metodoPago(PagoEntity.MetodoPago.TARJETA)
                .estadoPago(PagoEntity.EstadoPago.APROBADO)
                .referenciaTransaccion(referenciaTransaccion)
                .comprobanteUrl(comprobanteUrl)
                .fechaPago(LocalDateTime.now())
                .fechaCreacion(LocalDateTime.now())
                .fechaActualizacion(LocalDateTime.now())
                .build();

        PagoEntity pagoGuardado = pagoRepository.save(pago);
        
        // Actualizar venta a COMPLETADA
        venta.setEstado(VentaEntity.EstadoVenta.COMPLETADA);
        venta.setFechaActualizacion(LocalDateTime.now());
        ventaJpaRepository.save(venta);
        
        // Inicializar seguimiento logístico
        try {
            seguimientoLogisticaService.inicializarSeguimientoPorPagoAprobado(
                    venta.getId(), 
                    "Pago procesado con tarjeta: " + referenciaTransaccion
            );
            log.info("Seguimiento logístico inicializado para venta ID: {}", venta.getId());
        } catch (Exception e) {
            log.error("Error al inicializar seguimiento logístico para venta ID: {}", venta.getId(), e);
        }
        
        // 🧾 GENERAR FACTURA PDF DE FORMA ASÍNCRONA (no bloquea la respuesta)
        final Long ventaIdFinal = venta.getId();
        new Thread(() -> {
            try {
                Thread.sleep(100); // Pequeña pausa para asegurar que la transacción se complete
                String rutaFactura = facturaService.generarFacturaPDF(ventaIdFinal);
                log.info("✅ Factura generada en segundo plano: {}", rutaFactura);
            } catch (Exception e) {
                log.error("⚠️ Error al generar factura PDF para venta ID: {}", ventaIdFinal, e);
            }
        }).start();
        
        log.info("Pago con tarjeta procesado exitosamente. Pago ID: {}, Estado: APROBADO", pagoGuardado.getId());
        return pagoGuardado;
    }
}
